Documentation/design-docs/arm64-intrinsics.md

```diff
@@ -0,0 +1,217 @@
+  + All extensions which were not required when they were initially introduced
+  + For example, `CRC32` became required with `ARMv8.1`, but since it was introduced as optional in `ARMv8` it would go in its own namespace
+  
+If an extension is only supported on `Arm32`, there is no reason to add a sub namespace.
+
++ `System.Runtime.Intrinsics.Arm.ARMv7` is used for `ARMv7` added functionality not included in `Base`
+
+If an extension is only supported on `Arm64`, there is no reason to add a sub namespace.  Hypothetically:
+
++ `System.Runtime.Intrinsics.Arm.ARMv10` is used for `ARMv10` added functionality (Hypothetically `Arm64` only)
+
+Namespace is chosen based on what could be implemented, not on what is or will be implemented.
+
+#### Names
+
+Intrinsics will be named to describe functionality.  Names will not correspond to specific named assembly instructions.
+
+`IsSupported` will be used in any namespace which may not be implemented on all CoreCLR platforms
+
+### Intrinsic Interface Documentation
+
++ Documentation will be minimal preferring the underlying ARM documentation
++ Namespaces will briefly document corresponding specification
++ Intrinsic methods will briefly document corresponding assembly instruction(s) for each platform
+
+### Test coverage
+
+As intrinsic support is added test coverage must be extended to provide basic testing
+
+### Phased Implementation
+
+#### Implementation Priorities
+
+As guidelines:
+
++ Baseline functionality will be prioritized over architectural extensions
++ Architectural extensions will typically be prioritized in age order.  Earlier extensions will be added first
++ `ARM64` only and `ARM*` features will have equal priority
++ Priorities will be driven by optimization efforts and requests
++ Priority will be given to intrinsics which have already been implemented for other platforms
++ Priority will be given to intrinsics which are equivalent/similar to those actively used in libraries for other platforms
+
+#### Addition of new intrinsics namespaces
+When an intrinsic namespace is introduced code will likely be implemented on different platforms at different times.
+
+It is explicitly OK for ARM64 to implement a namespace which may be implemented on `ARM32` w/o `ARM32` implementation.
+The opposite is also true.
+
+Any platform which does not implement a namespace, must:
+
++ Return `false` for `IsSupported`
++ `throw` on any unimplemented function
+
+#### Partial implementation of intrinsic namespace
+
++ It is preferred that `IsSupported` represents the state of an entire namespace
++ It is certainly required at time of a release, `IsSupported` represents the state of an entire namespace
++ Allow for writing tests and implementing code, `IsSupported == false` may be used when namespace is partially implemented
+
+#### Addition of new intrinsics to existing namespace
+When an intrinsic namespace is introduced, a best guess will be made about the set of useful intrinsics.
```

Comment:
See remarks above - I think that best guess in the form of logical simulation was made by arm designers when they were designing ISA extensions. There is no reason to invalidate that work based on best guess.

Reply:
Best guess here is on the choice of assembly intrinsics to expose.  In general, I do not see it productive to automatically create an intrinsic for every assembly instruction.  Certainly 
+ `nop` doesn't need one
+ `brk` would be problematic
+ `mul` of general purpose register seems unnecessary

It may be the best guess means expose
+ Every vector instruction
+ Every crypto instruction
+ Every atomic instruction
+ Every floating point instruction

My guess is that may also be excessive.

Reply:
It seems we have an misunderstanding here. I have assumed that Arm intrinsics would be similar to Intel ones and would comprise only ISA extensions: SIMD, crypto etc.

Reply:
For ARM64, SIMD is not an extension it is a required part of the base instruction set.

There is no reason instrinsics have to be limited to an architectural extension.  The decision should be based on utility.  Every assembly instruction form is a candidate.  Some obviously do not make sense.  Others make sense when compared to current intrinsic decisions.  Others are unclear.

Reply:
It is possible for an AArch64 based machine to not support SIMD, however, that does not mean CoreCLR would run on it.
![image](https://user-images.githubusercontent.com/10487869/33567945-a6ca9788-d8d9-11e7-8d39-bff181d848eb.png)

That being said, I think this fits into the comment I made here: https://github.com/dotnet/coreclr/pull/15343#discussion_r154542336

If the support for a feature is visible via a CPU Register, then it should be in a class where IsSupported is a check against said register. This is similar to how all X64 processors ship with SSE/SSE2, but we still have separate classes for those intrinsics and the required checks.

Reply:
> It is possible for an AArch64 based machine to not support SIMD

I did not realize that option was added.

> If the support for a feature is visible via a CPU Register

These registers are not visible to user space apps.  ARM64 must rely on OS to expose information at whatever granularity it chooses.